### PR TITLE
Serialization#3

### PR DESCRIPTION
**Pull request by UraFIZ**
To make it possible to easily change the way of serializaition I have implemented the strategy pattern.
"Strategy is a behavioral design pattern that lets you define a family of algorithms, put each of them into a separate class, and make their objects interchangeable." - [link](https://refactoring.guru/design-patterns/strategy).

- family of algorithms - pickle and json
- interchangeability is achived by having the same name of methods

The implementation of the serializaition happens in high class Storag which behaives like a Fabric hiding the implementation. The BookManager inherits from the Storage and can manipulate its methods like save, load.

Note: to be able to implement json serialization I had to add methods are responsable for transform classe to the dictionary "from_dict" and vv "to_dict". That because the json library can't serializ classes.

### CODE REVIEW

**Comment by UraFIZ** [comment]
Gonna resolve conflicts and pull all staff together. Then ping guys to review.

**Comment by UraFIZ** [comment]
Here is how the saved data looks like

image

File: src/Storate.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+    def save(self, data: dict):
+        with open(self.file_name, self.write_modification) as file:
+            file.write(self.strategy.serialize(data))
+
+    def load(self) -> dict:
+        try:
+            with open(self.file_name, self.read_modificator) as file:
+                return self.strategy.deserialize(file.read())
+        except (self.error_exception):
```
Comment (nrkshaman) [inline src/Storate.py:20]: Тут ерори можна зробити в одному except
`except (self.error, FileNotFoundError):`
а якшо в except одна ерора то дужок не треба
І якшо добавиш туда EOFError (EndOfFileError) то ше покриєш випадок коли бінарник для пікла просто пустий

File: src/BookManager.py [hunk 2]
```diff
@@ -1,26 +1,44 @@
...
+import json
+
+from src.Storate import Storage
 from src.models.AddressBook import AddressBook
+from src.models.NoteBook import NoteBook
+from src.serialization.json_serialization_strategy import JsonSerializationStrategy
+from src.serialization.serialization_strategy import SerializationStrategy
 
 
-class BookManager:
```
Comment (nrkshaman) [inline src/BookManager.py:6]: В цьому класі в кількох місцях відступи не ок
Рядки 21,22,26,41

**Review by nrkshaman — APPROVED** [review]

File: src/Storate.py (hunk 1 above)
Comment (UraFIZ) [reply src/Storate.py]: Дякую!

**Pull request merged**